### PR TITLE
Uniffi top frecent site info

### LINKED ISSUE #4696: Uniffi Top Frecent Site Info Types
Uniffi the _TopFrecentSiteInfo_ and  _TopFrecentSiteInfos_ types (_FrecencyThresholdOption_ and _TopFrecentSiteInfo_ in kotlin) and the places_get_top_frecent_site_infos function.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2931)
┆Story Points: 3
┆epic: Uniffi Places
┆sprintEndDate: 2021-12-10

### PR DESCRIPTION
**Pull request by tarikeshaq**
fixes #4696 

Uniffies the top frecent site info types and function call

no change in behaviour it was a small mechanical change. Only thing to point out it the `url` is now represented as an actual `Url` type, and it would error out if it wasn't

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by tarikeshaq** [comment]
the swiftlint ci failure is fixed on main by https://github.com/mozilla/application-services/pull/4710 - it's unrelated

File: components/places/src/ffi.rs
Comment (skhamis) [inline components/places/src/ffi.rs:128]: I think we can move this into src/places/types.rs to keep things consistent

**Review by skhamis — COMMENTED** [review]
Awesome job with this! Will there be any PRs needed in a-c (not sure if this is implemented in iOS but also that?) Right now I have https://github.com/skhamis/android-components/tree/places-uniffication-ac that I've made just to be able to throw whatever relevant changes to tests/types into a simple PR just so we can make sure things are looking okay for the consumer.

**Comment by skhamis** [comment]
@tarikeshaq scratch the moving comment, didn't realize it was only the FFI type -- looks good! I'll approve if we can confirm no ios/android changes needed!

**Comment by lougeniaC64** [comment]
FYI This logic isn't used in iOS and probably won't affect android because the `TopFrecentSiteInfo` type moved from the kotlin layer to rust, however it might be worthwhile to do the same with the `FrecencyThresholdOption` type (which is in a-c and may require superficial code changes). This is a nit thought, just trying to get our mobile wrapper code as minimal as possible.

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt
Comment (mhammond) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt:313]: surprised this wasn't using the `readQueryCounters.measure` dance! But agree with not adding it as part of this PR

**Review by mhammond — APPROVED** [review]
Looks great!

**Comment by tarikeshaq** [comment]
Built against AC, it does break it, but the only change needed is https://github.com/tarikeshaq/android-components/commit/463624e532bf752c60c69130b968041211fc61c1#diff-bda6b853789a9b3c59a0579fbc6c852f6631a60f309fea114cf6e3b4964a6b8dR90, just adding a `.uniffi` makes it work 😄

**Comment by tarikeshaq** [comment]
@lougeniaC64 I took your feedback and moved the `FrecencyThresholdOption` into the Rust layer, again building against AC only requires an added `.uniffi` in the same file in AC

File: components/places/src/ffi.rs
Comment (tarikeshaq) [inline components/places/src/ffi.rs:146]: I'm a little iffy about those magic numbers... But they were magic numbers in the kotlin file before...

**Comment by lougeniaC64** [comment]
> @lougeniaC64 I took your feedback and moved the `FrecencyThresholdOption` into the Rust layer, again building against AC only requires an added `.uniffi` in the same file in AC

Awesome--thanks, @tarikeshaq!

File: components/places/src/ffi.rs
Comment (lougeniaC64) [reply components/places/src/ffi.rs:146]: I was hoping to find a useful comment [in desktop](https://searchfox.org/mozilla-central/source/browser/components/newtab/lib/TopSitesFeed.jsm#87), but that didn't pan out as I had hoped. 😬

Comment (tarikeshaq) [reply components/places/src/ffi.rs:146]: Lol I'll define the value as a constant at the top of the file, add some comments and a link to desktop at the very minimum, maybe that's enough for now 🤷

Comment (mhammond) [reply components/places/src/ffi.rs:146]: I *think* it relates to the constants [here](https://github.com/mozilla/application-services/blob/main/components/places/src/frecency.rs#L52,L59) and is basically saying "does this visit have something other than the smallest possible frecency?". Or something like that 😅

**Comment by tarikeshaq** [comment]
Created https://github.com/skhamis/android-components/pull/3 for the AC work associated with this PR, I'll land this PR and once I get feedback on that, I'll land it too

**Pull request merged**
